client/main.cpp

```diff
@@ -0,0 +1,176 @@
+          {{"username", "u"}, "The user name to use for the connection.", "username", ""},
+          {{"password", "P"}, "The password to use for the connection.", "password", ""},
+          {{"subscribe", "s"}, "Subscribe to a topic filter.", "topicfilter"},
+          {{"publish", "p"}, "Publish to topic.", "topic"},
+          {{"payload", "l"}, "Publish payload (requires -p).", "payload"},
+          {{"retain", "r"}, "Retain flag for publishes (default: no retaining)."},
+          {{"qos", "q"}, "QoS (default: 1).", "QoS", "1"},
+          {{"ssl", "S"}, "Use SSL for TCP connections (default: off) (Websocket connections will determine the use of SSL using the url scheme)."},
+          {{"accept-self-signed-certificate", "A"}, "Ignore self signed certificate errors"},
+          {{"verbose", "v"}, "Be more verbose"}
+      });
+    parser.addHelpOption();
+
+    parser.process(app->arguments());
+
+    if (parser.positionalArguments().count() < 1) {
+        parser.showHelp(-1);
+    }
+
+    QString verbose = parser.isSet("verbose") ? "true" : "false";
+    QLoggingCategory::setFilterRules(QString("nymea.mqtt.client.debug=%1\nnymea.mqtt.client.warning=%2\n").arg(verbose).arg(verbose));
+
+    bool ok;
+    Mqtt::QoS qos = static_cast<Mqtt::QoS>(parser.value("qos").toInt(&ok));
+    if (!ok) {
+        qCritical() << "Invalid QoS option. Options are 0, 1 or 2.";
+        exit(-1);
+    }
+    bool retain = parser.isSet("retain");
+
+    MqttClient client(parser.value("clientid"), app);
+    QObject::connect(&client, &MqttClient::error, app, [&client, app](QAbstractSocket::SocketError socketError){
+        qCritical() << "Connection error:" << socketError;
+        client.setAutoReconnect(false);
+        app->quit();
+    });
+    QObject::connect(&client, &MqttClient::sslErrors, app, [&client, &parser](const QList<QSslError> &sslErrors){
+        QSslCertificate certificate = sslErrors.first().certificate();
+        if (parser.isSet("accept-self-signed-certificate")) {
+            QList<QSslError> copy = sslErrors;
+            copy.removeOne(QSslError(QSslError::HostNameMismatch, certificate));
+            copy.removeAll(QSslError(QSslError::SelfSignedCertificate, certificate));
+            if (copy.isEmpty()) {
+                qInfo() << "Accepting self signed certificate";
+                client.ignoreSslErrors();
+                return;
+            }
+        }
+        qWarning() << "SSL errors for certificate:" << qUtf8Printable(certificate.toText()) << sslErrors;
+    });
+    QObject::connect(&client, &MqttClient::connected, app, [&parser, &client, qos, retain](Mqtt::ConnectReturnCode connectReturnCode, Mqtt::ConnackFlags /*connackFlags*/){
+        switch (connectReturnCode) {
+        case Mqtt::ConnectReturnCodeIdentifierRejected:
+            qCritical() << "Connection failed: Identifier rejected";
+            exit(-1);
```

Comment:
Not so important, only fyi: we actually have a constant for this called `EXIT_FAILURE` from stdlib return 1 and not -1.

Reply:
fixed